README.md
[![License: MPL 2.0][mpl-svg]][mpl]
[![Test Status][travis-badge]][travis]
[![Build Status][circleci-badge]][circleci]
[![Connect to Matrix via the Riot webapp][matrix-badge]][matrix]

# Pushbox - A rust implementation of Push/Sync long term storage

## What is it?

This is an internal project. mozilla needs the ability to store large data
chunks that may not fit perfectly within a standard WebPush message. PushBox
acts as an intermediary store for those chunks.

Messages are created by Firefox Accounts (FxA), stored here, and then a
WebPush message containing a URL that points back to this storage is sent
to the User Agent.

The User Agent can then fetch the data, decrypt it and do whatever it needs
to.

This project, once completed, will eventually replace the AWS Severless
PushBox project. It's being developed here because serverless can be a bit
greedy about what it grabs, and since PushBox is a rapid prototype, it's
good to treat it in a clean room environment.

See [API doc](
https://docs.google.com/document/d/1YT6gh125Tu03eM42Vb_LKjvgxc4qrGGZsty1_ajf2YM/)

## Requirements

The project requires Rust Nightly, a MySQL compliant data store, and
access to a [Firefox Accounts token verifier](https://github.com/mozilla/fxa-auth-server) system.


## Setting Up

1) Install Rust Nightly.

The rocket.rs [Getting Started](https://rocket.rs/guide/getting-started/)
document lists information on how to set that up.

2) create the pushbox MySQL user and database.

Because I'm horribly creative and because this is a WIP, I use "`test:test@localhost/pushbox`".
This is not recommended for production use. You can set your preferred
MySQL access credential information as "database_url" in the `Rocket.toml`
settings file (See [Rocket Config](https://rocket.rs/guide/configuration/#rockettoml)
information.)

3) Run `cargo run` to start the application, which (depending on the last
  commit) may actually start the program. YMMV.

## Running Docker Image

It is ***NOT*** advised to run the docker image in a production environment.
It's best suited for local development ***ONLY***.

Be sure to have `docker-compose` v. 1.21.0 or later installed.

1) From a dedicated screen (or tmux window)

$ `docker-compose up`

This will create two intertwined docker images:

***db_1*** - the database image. This image is a local test image. The database
can be accessed via `mysql -utest -ptest -h localhost --port 4306 pushbox`.

***app_1*** - the **pushbox** application. This is accessible via port
8000,
and uses the server authentication key of "[Correct_Horse_Battery_Staple_1](https://www.xkcd.com/936/)".

e.g.
` curl -H "Authorization: FxA-Server-Key Correct_Horse_Battery_Staple_1" "http://localhost:8000/v1/userid/deviceid"`

Note: No garbage collection is currently done for the database. Heavy use
might warrant deleting old records every so often.

# API

Pushbox is normally called via a HTTP interface using Authorized calls. Responses are generally JSON objects with appropriate HTTP status codes to indicate success/failure.

## Authorization

All calls to Pushbox require authorization. Authorization is specified by the `Authorization` header and can be either via a Server-Key or using Firefox Accounts OAuth token scopes. The method of authorization shall be determined and set by operations. It is **strongly** suggested that if Server-Key is used, access to PushBox be limited to an ACL of calling Sync servers. 

e.g.

```
Authorization: fxa-server-key Correct-Horse-Battery-Staple-1
```


## GET /v1/store/< user> /< device >[?< options >]

Fetch data for a `<user>` on a `<device>`. 

Options may be one or more of the following:

* *index* - offset index to being new messages.
* *limit* - maximum number of messages to return. This will include the next index value to use. 
* *status* - Quick set the index and limit values:
  * *new* - client is new, and needs all records.
  * *lost* - client only needs latest index.

The return value is JSON structure: 

```javascript
{ "last": true, /* boolean indicating this is the last data block */
  "index": 123, /* the highest message index value returned */
  "status": 200, /* HTTP status for result */
  "messages": [
    {"index": 123, /* Message block index */
     "data": "aBc1..." /* encrypted data block */
    }, ...
  ]
}
```

## POST /v1/store/< user >/< device >

Write a databblock for a `<user>` on `<device>`

The body of the post message is a JSON structure:

```javascript
{"ttl": 3600, /* Time for the data to live in seconds.*/
 "data": "aBc1..." /* Encrypted data block to store */
}
```

**NOTE:** Please be certain to encrypt the body of the data you wish to store. No encryption is done on the server side, and even if it was, there's no guarantee that it couldn't be reversed by a disgruntled employee or malicious agent.

The returned value is a JSON structure:

```javascript
{
  "status": 200, /* The HTTP status for the result */
  "index": 123 /* the index number of the stored record */
}
```

## DELETE /v1/store/< user >[/< device >]

Delete all records for a given user or just a given user's device.

This call returns just an empty object.


## GET /__heartbeat__

Return the status of the server.

This call is only used for server status checks.


## GET /__lbheartbeat__

Return a light weight status check (200 OK).

This call is only used for the Load Balancer's check.


## GET /__version__

Return a JSON response of the version information of the server.


# Database

Pushbox requires a configured MySQL compliant server. Pushbox requires the configuration file to contain the proper credentials, but will create any require table or indexes. Currently, there is no garbage collection done for expired, unread records. It is suggested that a regularly scheduled command be created to run 

```sql
DELETE from pushboxv1 where TTL < unix_timestamp();
```

This function may be added to pushbox at a later date.

[mpl-svg]: https://img.shields.io/badge/License-MPL%202.0-blue.svg
[mpl]: https://opensource.org/licenses/MPL-2.0
[travis-badge]: https://travis-ci.org/mozilla-services/pushbox.svg?branch=master
[travis]: https://travis-ci.org/mozilla-services/pushbox
[circleci-badge]: https://circleci.com/gh/mozilla-services/pushbox.svg?style=shield&circle-token=074ae89011d1a7601378c41a4351e1e03f1e8177
[circleci]: https://circleci.com/gh/mozilla-services/pushbox
[matrix-badge]: https://img.shields.io/badge/chat%20on%20[m]-%23push%3Amozilla.org-blue
[matrix]: https://chat.mozilla.org/#/room/#push:mozilla.org


CHANGELOG.md
<a name="0.3.2"></a>
## 0.3.2 (2021-09-24)

#### Bug Fixes

* fix db healthcheck [dc3a09922](https://github.com/mozilla-services/pushbox/commit/dc3a099220287bf3bbec942bd838dd0b8cfa4717), closes [#85](https://github.com/mozilla-services/pushbox/issues/85)
* include updates for 2021 Q3

<a name="0.3.0"></a>
## 0.3.0 (2020-02-29)


#### Refactor

*   cleanup no longer needed code ([a629a78c](https://github.com/mozilla-services/pushbox/commit/a629a78c6929c41750099f9e5b0283bd87d06b40))

#### Bug Fixes

*   fix query param handling per rocket 0.4 changes ([db6c36a7](https://github.com/mozilla-services/pushbox/commit/db6c36a769b33da5239f392ef58ab5dbd92c5256), closes [#61](https://github.com/mozilla-services/pushbox/issues/61))

#### Features

*   Differentiate GeneralError into more specific errors ([f328d882](https://github.com/mozilla-services/pushbox/commit/f328d882f6e568769b1148cde8b2ab4315766412), closes [#59](https://github.com/mozilla-services/pushbox/issues/59))

#### Chore

*   fix the mysql port ([3aca62a6](https://github.com/mozilla-services/pushbox/commit/3aca62a6eeeba62e0414ede3f67777006e19dd8e), closes [#65](https://github.com/mozilla-services/pushbox/issues/65))
*   cargo fix --edition-idioms ([51c1ea16](https://github.com/mozilla-services/pushbox/commit/51c1ea16f923090ea7eda931445565400ceb735c))
*   cargo fix --edition ([c7e2312f](https://github.com/mozilla-services/pushbox/commit/c7e2312fc868dea9292fe5f9639f15b39b82557d))
*   update deps ([733f6ac7](https://github.com/mozilla-services/pushbox/commit/733f6ac7cd6616bb21725fafae88c1180de4442c), closes [#63](https://github.com/mozilla-services/pushbox/issues/63))
*   update rust nightly ([6b24ecc8](https://github.com/mozilla-services/pushbox/commit/6b24ecc8654cc040ff82b651f20a2c7bcef92f22))
*   Dependency Update 10/2019 ([e9c1f043](https://github.com/mozilla-services/pushbox/commit/e9c1f043793d2ca60f444f713d2f8e935b668731))
*   Update deps ([0e426b0b](https://github.com/mozilla-services/pushbox/commit/0e426b0b332d18ea239bc34c09a0188a783b6332))
*   Update cargo deps ([40f3688a](https://github.com/mozilla-services/pushbox/commit/40f3688a78bc92fd2332c7016e5d7e8b586c2332))
*   Update dependencies ([b93d9183](https://github.com/mozilla-services/pushbox/commit/b93d9183cccac835912fbc0cb19cbf6cfe23ff35), closes [#53](https://github.com/mozilla-services/pushbox/issues/53))
*   Add Code of Conduct ([4b803d62](https://github.com/mozilla-services/pushbox/commit/4b803d6249568b4054a1df59a7db488852afb358))
*   library updates ([adad9ee4](https://github.com/mozilla-services/pushbox/commit/adad9ee40d9be32b2055cd845b74a82ccc31ee99))
*   remove /sls and add .dockerignore ([505bb197](https://github.com/mozilla-services/pushbox/commit/505bb19714de0d02fc22cf2eba15aba9080b90b1), closes [#46](https://github.com/mozilla-services/pushbox/issues/46))
*   utilize a rust-toolchain file ([2b65c02c](https://github.com/mozilla-services/pushbox/commit/2b65c02c42b77cacae9da61d3bb6cc1af98ac7b4))



<a name="0.1.2"></a>
## 0.1.2 (2018-08-21)


#### Bug Fixes

*   Fix deployment SQS issue ([c3ab1dc7](https://github.com/mozilla-services/pushbox/commit/c3ab1dc77f32b555f62fcf9ade428fc3c1d35a2e), closes [#42](https://github.com/mozilla-services/pushbox/issues/42))
*   Allow AWS_LOCAL_SQS to specify more than just custom regions. ([a51e6ef0](https://github.com/mozilla-services/pushbox/commit/a51e6ef0d2933652019a1a27050aede7e7baea7b))
*   skip sqs mod testing on travis ([b0f3508b](https://github.com/mozilla-services/pushbox/commit/b0f3508bd9a96a3114e039478e657dbbeee00188), closes [#39](https://github.com/mozilla-services/pushbox/issues/39))



<a name="0.1.1"></a>
## 0.1.1 (2018-07-31)


#### Bug Fixes

*   strip FxA-Request-Id from the response ([ad09cdaa](https://github.com/mozilla-services/pushbox/commit/ad09cdaace7236e012af4a2fef43c199afcee342), closes [#36](https://github.com/mozilla-services/pushbox/issues/36))

#### Features

*   hacky support of ROCKET_LOG=off ([1c63ee3a](https://github.com/mozilla-services/pushbox/commit/1c63ee3a0242ad7f90ab6a4a2280c70dfe2ecb0e), closes [#33](https://github.com/mozilla-services/pushbox/issues/33))



<a name="0.1.0"></a>
## 0.1.0 (2018-07-25)


#### Bug Fixes

*   fix aws paramater typing ([9130cd16](https://github.com/mozilla-services/pushbox/commit/9130cd16b303b6934d962e151912d50d8b0bdeb2))
*   unquote data response & fix limit=0 indexing (#19) ([ba491c9a](https://github.com/mozilla-services/pushbox/commit/ba491c9ae8712609baa7034087edeed66ae822bb))

#### Features

*   support dockerflow style health checks ([59949d5e](https://github.com/mozilla-services/pushbox/commit/59949d5ee0421c0a839e2ba4eff6268da33748cf), closes [#30](https://github.com/mozilla-services/pushbox/issues/30))
*   add pass thru trace id (#26) ([b8aeb628](https://github.com/mozilla-services/pushbox/commit/b8aeb628d6f2981b4b38b2d5e164d44e3f97d71b), closes [#24](https://github.com/mozilla-services/pushbox/issues/24))
*   Add version.json (#28) ([f502bacd](https://github.com/mozilla-services/pushbox/commit/f502bacdaefa8219a61f4e07c07ab0edc0de2948), closes [#27](https://github.com/mozilla-services/pushbox/issues/27))
*   add errnos ([b4f500e6](https://github.com/mozilla-services/pushbox/commit/b4f500e6b179a3680fb79842a2ed994b44798ea8), closes [#23](https://github.com/mozilla-services/pushbox/issues/23))
*   Add documentation (#22) ([402038ff](https://github.com/mozilla-services/pushbox/commit/402038ff4f442e394a17794b66ad2e5f1f4260ae))
*   support lcip.org domain ([11a5cd44](https://github.com/mozilla-services/pushbox/commit/11a5cd448e3174156f1574ed07479e6094b51e37))
*   Move auth to own handler ([a53029ae](https://github.com/mozilla-services/pushbox/commit/a53029aec62d79382f56dd08e975e0c3744cc5da))
*   add primitive delete function ([3bd4a359](https://github.com/mozilla-services/pushbox/commit/3bd4a3596d36a049bc76582cc48d86731e959cdf))

#### Chore


handler.py
import json
import logging
import os
import time
import uuid
from functools import wraps

import boto3
from boto3.dynamodb.conditions import Key, Attr
from botocore.exceptions import ClientError

# Logging
logger = logging.getLogger()
logger.setLevel(logging.INFO)

# Constants
DEFAULT_TTL = 60 * 60 * 24

# Environment Constants
S3_BUCKET = os.environ.get("S3_BUCKET", "pushbox-test")
DDB_TABLE = os.environ.get("DDB_TABLE", "pushbox_test")

# Clients
s3 = boto3.resource("s3")
ddb = boto3.resource("dynamodb")
index_table = ddb.Table(DDB_TABLE)


class HandlerException(Exception):
    def __init__(self, status_code=500, message="Unknown Error"):
        self.status_code = status_code,
        self.message = message

    def __str__(self):
        return "{}: {}".format(self.status_code, self.message)


def log_exceptions(f):
    @wraps(f)
    def wrapper(*args, **kwargs):
        try:
            return f(*args, **kwargs)
        except Exception as exc:
            logger.exception("Exception running _store_data: {}".format(exc))
            raise
    return wrapper


def compose_key(uid, device_id):
    return "{}:{}".format(uid, device_id)


def get_max_index(key):
    result = index_table.query(
        KeyConditionExpression=Key("fxa_uid").eq(key),
        Select="ALL_ATTRIBUTES",
        ScanIndexForward=False,
        Limit=1,
    )
    if result['Count']:
        return int(result['Items'][0].get('index'))
    else:
        return 0


@log_exceptions
def store_data(event, context):
    """Store data in S3 and index it in DynamoDB"""
    logger.info("Event: {} ; Context: {}".format(event, context))
    device_id = event["pathParameters"]["deviceId"]
    fx_uid = event["pathParameters"]["uid"]
    try:
        key = compose_key(uid=fx_uid, device_id=device_id)
        req_json = json.loads(event["body"])
        logger.info("data: {}".format(req_json))
    except ValueError as ex:
        return dict(
            headers={"Content-Type": "application/json"},
            statusCode=400,
            body=json.dumps(dict(
                status=400,
                error="Invalid payload: {}".format(ex)
            ))
        )
    ttl = req_json.get("ttl", DEFAULT_TTL)
    s3_filename = device_id + uuid.uuid4().hex
    # Data is presumed to be an encrypted blob
    s3.Object(S3_BUCKET, s3_filename).put(Body=req_json["data"])
    index = get_max_index(key) + 1
    data_len = len(req_json["data"])
    for i in range(0, 10):
        try:
            index = index + i
            index_table.put_item(
                ConditionExpression=Attr('index').ne(index),
                Item=dict(
                    fxa_uid=key,
                    index=index,
                    device_id=device_id,
                    ttl=int(time.time()) + ttl,
                    s3_filename=s3_filename,
                    s3_file_size=data_len
                )
            )
            break
        except ClientError as ex:
            if ex.response['Error']['Code'] == \
                    'ConditionalCheckFailedException':
                pass
    return dict(
        headers={"Content-Type": "application/json"},
        statusCode=200,
        body=json.dumps(dict(
            status=200,
            index=index,
        ))
    )


@log_exceptions
def get_data(event, context):
    """Retrieve data from S3 using DynamoDB index"""
    logger.info("Getting data: {}".format(event))
    device_id = event["pathParameters"]["deviceId"]
    fx_uid = event["pathParameters"]["uid"]
    limit = None
    # event could set "queryStringParameters" to the value None
    params = event.get("queryStringParameters")
    if params is not None:
        try:
            limit = int(params.get("limit"))
        except (ValueError, TypeError):
            limit = None
    if limit is None:
        limit = 10
    limit = min(10, max(0, limit))
    key = compose_key(uid=fx_uid, device_id=device_id)
    if limit == 0:
        index = get_max_index(key)
        return dict(
            headers={"Content-Type": "application/json"},
            statusCode=200,
            body=json.dumps(dict(index=index))
        )
    start_index = None
    if "index" in (event.get("queryStringParameters") or {}):
        start_index = int(event["queryStringParameters"]["index"])
        logger.info("Start index: {}".format(start_index))
    key_cond = Key("fxa_uid").eq(key)
    if start_index:
        key_cond = key_cond & Key("index").gt(start_index)
    results = index_table.query(
        Select="ALL_ATTRIBUTES",
        KeyConditionExpression=key_cond,
        ConsistentRead=True,
        Limit=limit,
    ).get("Items", [])
    logger.info("results: {}".format(results))
    # Fetch all the payloads
    for item in results:
        try:
            response = s3.Object(S3_BUCKET, item["s3_filename"]).get()
            data = response["Body"].read().decode('utf-8')
            item["data"] = data
        except ClientError as ex:
            logger.error(ex)
            return dict(
                headers={"Content-Type": "application/json"},
                statusCode=404,
                body="Content missing or expired"
            )
    # Serialize the results for delivery
    index = 0
    messages = []
    for result in results:
        result_index = int(result['index'])
        messages.append({'index': result_index,
                         'data': result['data']})
        index = max(index, result_index)
    payload = {"last": True, "index": start_index or 0}
    if results:
        # should this be comparing against "scannedCount"?
        payload["last"] = len(results) < limit
        payload["index"] = index
        payload["messages"] = messages
    return dict(
        headers={"Content-Type": "application/json"},
        statusCode=200,
        body=json.dumps(payload),
    )


@log_exceptions
def del_data(event, context=None):
    """Delete data for a given user/device/channel"""
    uid = event['pathParameters']["uid"]
    device_id = event['pathParameters']['deviceId']
    key = compose_key(uid=uid, device_id=device_id)
    items = index_table.query(
        Select="ALL_ATTRIBUTES",
        KeyConditionExpression=Key("fxa_uid").eq(key),
        ConsistentRead=True,
    ).get("Items", [])
    try:
        for item in items:
            s3.Object(S3_BUCKET, item["s3_filename"]).delete()
    except Exception as ex:
        return dict(
            headers={"Content-Type": "application/json"},
            statusCode=500,
            body=json.dumps(dict(
                status=500,
                error="Could not delete all data {}".format(ex),
            ))
        )
    return dict(
        headers={"Content-Type": "application/json"},
        statusCode=200,
        body="{}"
    )


@log_exceptions
def status(event, context=None):
    # Faux status to discover root.
    return dict(
        headers={"Content-Type": "application/json"},
        statusCode=200,
        body=json.dumps(dict(status=200, message="ok"))
    )


def test_index_storage():
    data = "BlockOfEncryptedStuff"
    store_result = store_data({
        "pathParameters": {
            "deviceId": "device-123",
            "uid": "uid-123"
        },
        "body": json.dumps({
            "data": data
        })
    }, None)
    fetch_result = get_data({
        "pathParameters": {
            "deviceId": "device-123",
            "uid": "uid-123"
        },
        "queryStringParameters": {
            "index": json.loads(store_result['body'])['index'] - 1
        },
    }, None)
    body = json.loads(fetch_result['body'])
    high_index = body['index']
    assert(body['last'])
    assert(body['index'] == json.loads(store_result['body'])['index'])
    assert(body['messages'][0]['data'] == data)

    index = get_data(
        {
            "pathParameters": {
                "deviceId": "device-123",
                "uid": "uid-123"
            },
            "queryStringParameters": {
                "limit": "0"
            },
        }, None)
    assert(high_index == json.loads(index["body"])["index"])
    print('Ok')


def test_delete_storge():
    del_data({
        "pathParameters": {
            "deviceId": "device-123",
            "uid": "uid-123"
        },
    })
    print("Ok")


if __name__ == "__main__":
    print("testing indexed data storage...")
    test_index_storage()
    print("testing delete...")
    test_delete_storge()
